### PR TITLE
[logs] Tail docker containers from file

### PR DESCRIPTION
**Pull request by prognant**
### What does this PR do?
Add the ability to tail docker container from file.
Introduce an option to choose between docker socket and file tailing (`logs_config.docker_container_use_file `)

### Motivation

Bypass the docker socket to increase the performance and lessen the
pressure on the docker daemon.

### Additional Notes
Doc to be updated as the agent would need to have read access to /var/lib/docker/containers/... :
* The volume should be mounted inside the containerised agent (general case, ECS included)
* The agent (DEB/RPM packaged deployment) may be running in the same env as the docker daemon and some permission adjustment may allow this feature to be used in that case
 
This feature is disabled by default as enabling it may cause registry offset upon upgrade. This may be addressed later.
This feature only works on linux, windows support will come in a subsequent PR.

See the RFC for extra details.
JIRA ref ac-614

Status output for `container_collect_all` source:
```
  container_collect_all
  ---------------------
    - Type: docker
      Status: OK
      BytesRead: 1278
      Container Info:
        Container ID: bb9550541655, Image: flog, Created: 2021-01-06T17:42:22.51003287Z, Tailing from file: /var/lib/docker/containers/bb9550541655aec0f7d11566a7d1a6f29c32d774c5906aa390fc98cea41eb520/bb9550541655aec0f7d11566a7d1a6f29c32d774c5906aa390fc98cea41eb520-json.log
        Container ID: fcbf622d31ef, Image: flog, Created: 2021-01-06T17:42:27.375435144Z, Tailing from file: /var/lib/docker/containers/fcbf622d31ef69973013e78db8b1d3a330b76132cb8c1e261930b1d61987840b/fcbf622d31ef69973013e78db8b1d3a330b76132cb8c1e261930b1d61987840b-json.log
    - Type: file
      Identifier: bb9550541655aec0f7d11566a7d1a6f29c32d774c5906aa390fc98cea41eb520
      Path: /var/lib/docker/containers/bb9550541655aec0f7d11566a7d1a6f29c32d774c5906aa390fc98cea41eb520/bb9550541655aec0f7d11566a7d1a6f29c32d774c5906aa390fc98cea41eb520-json.log
      Status: OK
      Inputs: /var/lib/docker/containers/bb9550541655aec0f7d11566a7d1a6f29c32d774c5906aa390fc98cea41eb520/bb9550541655aec0f7d11566a7d1a6f29c32d774c5906aa390fc98cea41eb520-json.log
      BytesRead: 714
    - Type: file
      Identifier: fcbf622d31ef69973013e78db8b1d3a330b76132cb8c1e261930b1d61987840b
      Path: /var/lib/docker/containers/fcbf622d31ef69973013e78db8b1d3a330b76132cb8c1e261930b1d61987840b/fcbf622d31ef69973013e78db8b1d3a330b76132cb8c1e261930b1d61987840b-json.log
      Status: OK
      Inputs: /var/lib/docker/containers/fcbf622d31ef69973013e78db8b1d3a330b76132cb8c1e261930b1d61987840b/fcbf622d31ef69973013e78db8b1d3a330b76132cb8c1e261930b1d61987840b-json.log
      BytesRead: 564
```

Status output for other sources:
```
  docker
  ------
    - Type: docker
      Status: OK
      BytesRead: 37264
      Container Info:
        Container ID: 26ff2a07e094, Image: flog, Created: 2021-01-06T16:18:10.602282008Z, Tailing from file: /var/lib/docker/containers/26ff2a07e094e6864aff043c6245a060c97339a2c01a505ef189582ba211dccb/26ff2a07e094e6864aff043c6245a060c97339a2c01a505ef189582ba211dccb-json.log
    - Type: file
      Identifier: 26ff2a07e094e6864aff043c6245a060c97339a2c01a505ef189582ba211dccb
      Path: /var/lib/docker/containers/26ff2a07e094e6864aff043c6245a060c97339a2c01a505ef189582ba211dccb/26ff2a07e094e6864aff043c6245a060c97339a2c01a505ef189582ba211dccb-json.log
      Status: OK
      Inputs: /var/lib/docker/containers/26ff2a07e094e6864aff043c6245a060c97339a2c01a505ef189582ba211dccb/26ff2a07e094e6864aff043c6245a060c97339a2c01a505ef189582ba211dccb-json.log
      BytesRead: 37264
```
### Describe your test plan
* Test plan to cover before QA cycle https://gist.github.com/prognant/5a25608c93c1470d554a6aeb6728ecdc
* Pieces of logs agent QA board

### CODE REVIEW

File: pkg/logs/input/docker/launcher.go [hunk 1]
```diff
@@ -192,6 +212,52 @@ func (l *Launcher) overrideSource(container *Container, source *config.LogSource
...
+
+	standardService := l.serviceNameFunc(container.container.Name, dockerutil.ContainerIDToTaggerEntityName(containerID))
+	shortName, err := container.getShortImageName()
+
+	if err != nil {
+		log.Warnf("Could not get short image name for container %v: %v", ShortContainerID(containerID), err)
+	}
+
+	// If collect all may be update info for consistency with docker socket tailing
+	// source.UpdateInfo(containerID, fmt.Sprintf("Container ID: %s, Image: %s, Created: %s", ShortContainerID(containerID), shortName, container.container.Created))
```
Comment (gh123man) [inline pkg/logs/input/docker/launcher.go:232]: Should this be uncommented?

Comment (prognant) [reply pkg/logs/input/docker/launcher.go:232]: I need to work on harmonization when tailing from file vs. when tailing from the Docker socket. Especially given the fact that tailing from file is done by emitting a new source to reuse the existing file logic. And that new source is appearing by itself in the status page. I don't think I will touch much of the tailing from socket execution path, but I think I will reuse the parent source logic you introduce to link the source create for the file and use that link to build a consistent status.

File: pkg/logs/input/container/launcher.go [hunk 2]
```diff
@@ -26,28 +26,28 @@ import (
 // If none of those volumes are mounted, returns a lazy docker launcher with a retrier to handle the cases
 // where docker is started after the agent.
 // dockerReadTimeout is a configurable read timeout for the docker client.
-func NewLauncher(collectAll bool, collectFromFiles bool, dockerReadTimeout time.Duration, sources *config.LogSources, services *service.Services, pipelineProvider pipeline.Provider, registry auditor.Registry) restart.Restartable {
+func NewLauncher(collectAll bool, kubernetesCollectFromFiles bool, dockerCollectFromFiles bool, dockerReadTimeout time.Duration, sources *config.LogSources, services *service.Services, pipelineProvider pipeline.Provider, registry auditor.Registry) restart.Restartable {
```
Comment (gh123man) [inline pkg/logs/input/container/launcher.go:29]: I think we should update the method doc to reflect the new parameters. ex:
```
// collectFromFiles is enabled.
```
is no longer accurate.

File: pkg/logs/input/docker/launcher.go [hunk 3]
```diff
@@ -212,19 +281,49 @@ func newOverridenSource(standardService, shortName string, status *config.LogSta
...
 	containerID := container.service.Identifier
-	if _, isTailed := l.getTailer(containerID); isTailed {
+	if _, isTailed := l.fileSourcesByContainer[containerID]; isTailed {
 		log.Warnf("Can't tail twice the same container: %v", ShortContainerID(containerID))
 		return
 	}
-
 	// overridenSource == source if the containerCollectAll option is not activated or the container has AD labels
-	overridenSource := l.overrideSource(container, source)
+	fileSource := l.getFileSource(container, source)
```
Comment (gh123man) [inline pkg/logs/input/docker/launcher.go:299]: I think this comment needs to be updated, `overridenSource` does not exist in this context

**Review by gh123man — APPROVED** [review]
LGTM. Just some minor comments to update comments.

**Comment by prognant** [comment]
Minor update: meaningful error when the docker directory is not readable and explicitly ignore windows platform (support is coming later).

**Comment by prognant** [comment]
Did another round of tests on ECS, the feature is working as expected:
```
  container_collect_all
  ---------------------
    - Type: docker
      Status: OK
      BytesRead: 90424
      Container Info:
        Container ID: 21b06c651137, Image: prognant, Created: 2021-01-20T14:14:45.551064312Z, Tailing from file: /var/lib/docker/containers/21b06c651137f8e8e7d43426a9169c7792589bdebadf2d5169f3681c1706ad6c/21b06c651137f8e8e7d43426a9169c7792589bdebadf2d5169f3681c1706ad6c-json.log
        Container ID: 0e8b8a15e6eb, Image: flog, Created: 2021-01-20T14:14:41.416343523Z, Tailing from file: /var/lib/docker/containers/0e8b8a15e6eb1c95be1c57bec195aa33bd06370633d0a6f279ab13023726d04e/0e8b8a15e6eb1c95be1c57bec195aa33bd06370633d0a6f279ab13023726d04e-json.log
        Container ID: 53ad82507959, Image: amazon-ecs-agent, Created: 2020-10-29T06:42:44.574368897Z, Tailing from file: /var/lib/docker/containers/53ad82507959249199f34a96e96921b1c93114f548fb2732ab0754f1480bd904/53ad82507959249199f34a96e96921b1c93114f548fb2732ab0754f1480bd904-json.log
    - Type: file
      Identifier: 21b06c651137f8e8e7d43426a9169c7792589bdebadf2d5169f3681c1706ad6c
      Path: /var/lib/docker/containers/21b06c651137f8e8e7d43426a9169c7792589bdebadf2d5169f3681c1706ad6c/21b06c651137f8e8e7d43426a9169c7792589bdebadf2d5169f3681c1706ad6c-json.log
      Status: OK
      Inputs: /var/lib/docker/containers/21b06c651137f8e8e7d43426a9169c7792589bdebadf2d5169f3681c1706ad6c/21b06c651137f8e8e7d43426a9169c7792589bdebadf2d5169f3681c1706ad6c-json.log
      BytesRead: 44739
    - Type: file
      Identifier: 0e8b8a15e6eb1c95be1c57bec195aa33bd06370633d0a6f279ab13023726d04e
      Path: /var/lib/docker/containers/0e8b8a15e6eb1c95be1c57bec195aa33bd06370633d0a6f279ab13023726d04e/0e8b8a15e6eb1c95be1c57bec195aa33bd06370633d0a6f279ab13023726d04e-json.log
      Status: OK
      Inputs: /var/lib/docker/containers/0e8b8a15e6eb1c95be1c57bec195aa33bd06370633d0a6f279ab13023726d04e/0e8b8a15e6eb1c95be1c57bec195aa33bd06370633d0a6f279ab13023726d04e-json.log
      BytesRead: 1592
    - Type: file
      Identifier: 53ad82507959249199f34a96e96921b1c93114f548fb2732ab0754f1480bd904
      Path: /var/lib/docker/containers/53ad82507959249199f34a96e96921b1c93114f548fb2732ab0754f1480bd904/53ad82507959249199f34a96e96921b1c93114f548fb2732ab0754f1480bd904-json.log
      Status: OK
      Inputs: /var/lib/docker/containers/53ad82507959249199f34a96e96921b1c93114f548fb2732ab0754f1480bd904/53ad82507959249199f34a96e96921b1c93114f548fb2732ab0754f1480bd904-json.log
      BytesRead: 44093
```

**Pull request merged**